### PR TITLE
Ane-1711 jar fingerprints testing [3/3]

### PR DESCRIPTION
**Pull request by csasarak**
# Overview

This adds a little automated testing around  analyzing jars in containers with millhone.

It also modifies the way we look up filepaths in tar files to normalize them. We avoided this in the past because I guess most of the tarballs we were scanning had everything at the root level so filepath separators didn't cause an error. This isn't true for newer images I've made with `docker save`. 

## Acceptance criteria

There are now automated tests that exercising reading jars out of containers with millhone and processing them in the CLI. 

## Testing plan

This PR is just tests to run. Manual tests have been done on the branches this is stacked on top of. 

## Risks

None.

## Metrics

_Is this change something that can or should be tracked? If so, can we do it today? And how? If its easy, do it_

## References

- _[ANE-1711](https://fossa.atlassian.net/browse/ANE-1711)_

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] If this PR added docs, I added links as appropriate to the user manual's ToC in `docs/README.ms` and gave consideration to how discoverable or not my documentation is.
- [x] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json` AND I have updated example files used by `fossa init` command. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).
- [x] If I made changes to a subcommand's options, I updated `docs/references/subcommands/<subcommand>.md`.

[ANE-1711]: https://fossa.atlassian.net/browse/ANE-1711?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: extlib/millhone/src/cmd/analyze_container.rs [hunk 1]
```diff
@@ -185,3 +185,54 @@ fn buffer(mut reader: impl Read) -> Result<Vec<u8>> {
...
+    }
+
+    #[test]
+    fn it_finds_expected_output() {
+        let image_tar_file =
+            PathBuf::from("../../test/App/Fossa/Container/testdata/jar_test_container.tar");
+
+        let res = jars_in_container(&image_tar_file).expect("Read jars out of container image.");
+
+        assert_eq!(res, expected_jar_analysis());
```
Comment (jssblck) [inline extlib/millhone/src/cmd/analyze_container.rs:236]: This is _fine_, but I **strongly** recommend not using the built-in types for this!

If you use the actual type in the test harness that the application uses, and the type changes its serialization, your test won't notice that you've broken backwards compatibility.

I strongly recommend instead testing against the raw JSON value:
```rust
let expected = serde_json::json!({
  "discovered_jars": {
    "blobs/sha256/61aed1a8baa251dee118b9ab203c1e420f0eda0a9b3f9322d67d235dd27a12ee": [
      {
        "kind": "v1.discover.binary.jar",
        "path": "jackson-annotations-2.17.1.jar",
        "fingerprints": {
          "v1.raw.jar": "wjGJk8cvY4tpKcUC5r8YuO15Wfv+rVuyWANBYCUIeDs=",
          "v1.class.jar": "t2Btr6rNrvzghM5Nud2uldRGVjw0/n5rK9j0xooQQyk=",
          "sha_256": "/MrYLhMXLA5DhNtxV3IZybhjHAgg9LGNqqVwFvtmHHY=",
          "v1.mavencentral.jar": "/KfvYZLJrQXQe8UNqZG/k3qErzo="
        }
      }
    ],
    "blobs/sha256/4d84019f77d1aa837a2cb4225bb79fc03a45ae5a247284dda07cfb9fb8077bd1": [
      {
        "kind": "v1.discover.binary.jar",
        "path": "inner_directory/commons-email2-jakarta-2.0.0-M1.jar",
        "fingerprints": {
          "sha_256": "MuEcK3nOFuySTTg4HOJi3vvTpI9bYspfMHa9AK2merQ=",
          "v1.mavencentral.jar": "6bzpyKql6Q+UxKQgm14pcP4wHGo=",
          "v1.class.jar": "2wRGbMGyGRwEXqNm53h1YK8OO879kvzDxazmJXiAcfI=",
          "v1.raw.jar": "QA4SAurtJeo+lx1Vqve5uQYnvDKLFx1NgsSuoWKi8pw="
        }
      }
    ],
    "blobs/sha256/cc2447e1835a40530975ab80bb1f872fbab0f2a0faecf2ab16fbbb89b3589438": []
  }
});

let image_tar_file = PathBuf::from("../../test/App/Fossa/Container/testdata/jar_test_container.tar");
let res = jars_in_container(&image_tar_file).expect("Read jars out of container image");
let res = serde_json::to_value(res).expect("encode as json");

// Optional, but `pretty_assertions` makes for dramatically better test failures
pretty_assertions::assert_eq!(expected, res);
```

**Review by jssblck — APPROVED** [review]

File: extlib/millhone/src/cmd/analyze_container.rs (hunk 1 above)
Comment (csasarak) [reply extlib/millhone/src/cmd/analyze_container.rs]: Thanks, that was a good call out. I've changed the test to do this.

File: extlib/millhone/src/cmd/analyze_container.rs [hunk 2]
```diff
@@ -24,17 +24,17 @@ pub struct Subcommand {
...
 const JAR_OBSERVATION: &str = "v1.discover.binary.jar";
 
-#[derive(Debug, PartialEq, Eq, Serialize, Clone)]
+#[derive(Debug, PartialEq, Eq, Deserialize, Serialize, Clone)]
 struct DiscoveredJar {
-    kind: &'static str,
+    kind: String,
     path: PathBuf,
     fingerprints: Combined,
 }
```
Comment (jssblck) [inline extlib/millhone/src/cmd/analyze_container.rs:32]: Since we no longer are deserializing this type, we should revert the `Deserialize` changes imo

**Review by jssblck — APPROVED** [review]

File: extlib/millhone/src/cmd/analyze_container.rs (hunk 2 above)
Comment (csasarak) [reply extlib/millhone/src/cmd/analyze_container.rs]: Done, thanks!

**Pull request merged**